### PR TITLE
Pickup latest release from cJSON

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

Pick up change from https://github.com/DaveGamble/cJSON/releases/tag/v1.7.16

*Description of changes:*

- Copy/pasted from the cJSON repo, and updated based on our lint.
- To review, you can ignore the white space.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/external/cJSON.c [hunk 1]
```diff
@@ -1028,954 +1038,959 @@ static cJSON_bool print_object(const cJSON * const item, printbuffer * const out
...
-  }
-
-  return item;
+    parse_buffer buffer = { 0, 0, 0, 0, { 0, 0, 0 } };
+    cJSON *item = NULL;
+
+    /* reset error position */
+    global_error.json = NULL;
+    global_error.position = 0;
+
```
Comment (waahm7) [inline source/external/cJSON.c:1104]: global_error is removed in our copy of cJson due to race conditions https://github.com/awslabs/aws-c-common/pull/972/files

File: source/external/cJSON.c [hunk 2]
```diff
@@ -86,87 +86,93 @@ THE SOFTWARE.
...
-  size_t position;
+    const unsigned char *json;
+    size_t position;
 } error;
+static error global_error = { NULL, 0 };
+
+CJSON_PUBLIC(const char *) cJSON_GetErrorPtr(void)
+{
+    return (const char*) (global_error.json + global_error.position);
+}
```
Comment (waahm7) [inline source/external/cJSON.c]: global_error and `cJSON_GetErrorPtr` is removed in our copy of cJson due to race conditions [#972 (files)](https://github.com/awslabs/aws-c-common/pull/972/files)

File: source/external/cJSON.c [hunk 3]
```diff
@@ -177,844 +183,848 @@ static void * CJSON_CDECL internal_realloc(void *pointer, size_t size)
...
+
+    /* This checks for NaN and Infinity */
+    if (isnan(d) || isinf(d))
+    {
+        length = snprintf((char*)number_buffer, sizeof(number_buffer) / sizeof(char), "null");
+    }
+	else if (d == (double)item->valueint) // NOLINT
+	{
+        length = snprintf((char*)number_buffer, sizeof(number_buffer) / sizeof(char), "%d", item->valueint);
+	}
```
Comment (waahm7) [inline source/external/cJSON.c:563]: bad lint? Not sure why we haven't removed NOLINT in this case

**Review by waahm7 — APPROVED** [review]
Fix & Ship. I think we should just disable lint on these files and keep it similar to external cJson as much as possible.

File: include/aws/common/external/cJSON.h [hunk 4]
```diff
@@ -280,15 +280,22 @@ CJSON_PUBLIC(cJSON*) cJSON_AddObjectToObject(cJSON * const object, const char *
 CJSON_PUBLIC(cJSON*) cJSON_AddArrayToObject(cJSON * const object, const char * const name);
 
 /* When assigning an integer value, it needs to be propagated to valuedouble too. */
-#define cJSON_SetIntValue(object, number) ((object) ? (object)->valueint = (object)->valuedouble = (number) : (number)) //NOLINT
+#define cJSON_SetIntValue(object, number) ((object) ? (object)->valueint = (object)->valuedouble = (number) : (number))
 /* helper for the cJSON_SetNumberValue macro */
 CJSON_PUBLIC(double) cJSON_SetNumberHelper(cJSON *object, double number);
-#define cJSON_SetNumberValue(object, number) ((object != NULL) ? cJSON_SetNumberHelper(object, (double)number) : (number)) //NOLINT
+#define cJSON_SetNumberValue(object, number) (((object) != NULL) ? cJSON_SetNumberHelper((object), (double)(number)) : (number))
```
Comment (waahm7) [inline include/aws/common/external/cJSON.h:286]: Brackets arount object seems redundant 🤔 Ignore if there is a reason for those
```suggestion
#define cJSON_SetNumberValue(object, number) ((object != NULL) ? cJSON_SetNumberHelper(object, (double)(number)) : (number))
```

File: source/external/cJSON.c (hunk 3 above)
Comment (TingDaoK) [reply source/external/cJSON.c]: aha, it's white space vs tab.

File: include/aws/common/external/cJSON.h (hunk 4 above)
Comment (waahm7) [reply include/aws/common/external/cJSON.h:286]: Saw the lint errors, we should probably just disable lint on these files.

File: include/aws/common/external/cJSON.h [hunk 5]
```diff
@@ -28,10 +28,10 @@ THE SOFTWARE.
  *   (3) Remove cJSON_GetErrorPtr and global_error as they are not thread-safe.
  */
 
-/* clang-format off */
+/* NOLINTBEGIN */
```
Comment (DmitriyMusatkin) [inline include/aws/common/external/cJSON.h:32]: why are we reenabling clang format? if we disabling tidy, we should disable format as well

Comment (TingDaoK) [reply include/aws/common/external/cJSON.h:32]: there is a .clang-format config that disabled the format already.

**Pull request merged**